### PR TITLE
Raiden now chooses the PFS from a Service Registry smart contract

### LINKED ISSUE #3467: User Policy - Raiden needs to choose a pathfinding service from ServiceRegistry
## Introduction

Right now, Raiden will start using PFS once the `--path-finding-service` flag is provided. However, the PFS should be (for now) randomly selected from the user registry

## Tasks
- [ ] Remove the flag
- [ ] Query the service registry contract for the list PF service providers
- [ ] Randomly select a PF service provider and use it

Depends on #3581 

Expected: End of Feb, 2019.

### PR DESCRIPTION
**Pull request by LefterisJP**
Fix #3467 

## What this PR does

- As the `ServiceRegistry` is really not well defined and a lot of open questions remain, this PR is just the first step of integrating with the contract as it is now in the smart contracts repo.
- Makes usage of PFS as the default
- If a `--pfs-address` is provided then that PFS is used
- If a `--pfs-address` is not provided then Raiden tries to get a random PFS address from the service registry.
- If the user provides the `--use-basic-routing` argument we revert to the old routing behavior.

### Service Registry specification
The `ServiceRegistry` is not at all well defined at the moment. We got many open questions. How many Service Registries will exist, what will the interface be, how can people deregister from the registry e.t.c. Check https://github.com/raiden-network/team/issues/363 for more details.

Considering the above, this PR simply performs the first step of integrating a simple proxy for the `ServiceRegistry` and assuming there is only one `ServiceRegistry` in existence for now, just to enable the bare bones of the envisioned functionality and enable testing and integration with the PFS.

### Default routing functionality

I changed the default being to use the PFS (either via the address argument, or the Registry). If the user wants to use the old basic routing he now needs to provide a special argument.

I believe this is the arguments we would want when we go to production with PFS. This may not be desired so early in the development stage so I can change the behaviour.

Alternative approach would be:

- If user provides pfs address use that.
- Add a new boolean argument `--use-service-registry`. If given then Raiden will pick a pfs from the service registry.
- If nothing is provided we just use the old routing.

On the other hand ... people who use master should just be prepared for changes in defaults. I am open to suggestions here.

### CODE REVIEW

**Comment by palango** [comment]
> I believe this is the arguments we would want when we go to production with PFS. This may not be desired so early in the development stage so I can change the behaviour.

I'd say this depends on whether or not we want to do another mainnet bugfix release (which I would argue for). In that case I think we should keep the PFS usage optional for now.

**Comment by ulope** [comment]
IMO the CLI interface for the PFS is becoming too confusing.

We'll now have three options that influence the way routing is done:
- `--use-basic-routing` - disables PFS entirely
- `--service-registry-contract-address` - default in dev, makes raiden chose a random PFS
- `--pathfinding-service-address` - overrides the service registry but only **if** `--use-basic-routing` is not given.

I'd suggest to change this to:
- `--routing-mode`, choice of:
  - `basic`, *default in prod mode* - use local routing
  - `pfs`, *default in dev mode* - use pfs
- `--pathfinding-service-address`, valid options:
  - `auto`, *default* - chose a pfs from the service registry if routing mode is set to `pfs`
  - `<URL>` - give explicit URL

**Comment by LefterisJP** [comment]
@ulope I like your suggesion better than my alternative. Thank you. I will implement this.

File: raiden/network/pathfinding.py [hunk 1]
```diff
@@ -1,19 +1,107 @@
...
+
+    If pfs_address is None then basic routing must have been requested.
+    If pfs_address is provided we use that.
+    If pfs_address is 'auto' then we randomly choose a PFS address from the registry
+
+    Returns the PFS url to use or None if we don't use the PFS and use basic routing
+    """
+    if routing_mode == RoutingMode.BASIC:
+        click.secho(
+            'Requested to use basic routing. No path finding service is being used',
```
Comment (ulope) [inline raiden/network/pathfinding.py]: This should also be logged (in order for e.g. debug logs to show the complete state).

And a nitpick: The phrasing sounds slightly awkward. Suggestion: `Not using path finding services, falling back to basic routing.`

File: raiden/network/pathfinding.py [hunk 2]
```diff
@@ -1,19 +1,107 @@
...
+        )
+        return None
+
+    msg = 'With PFS routing mode we shouldnt get to configure pfs with pfs_address being None'
```
Comment (ulope) [inline raiden/network/pathfinding.py]: Maximum nitpick: `shouldnt` -> `shouldn't`

File: raiden/network/pathfinding.py [hunk 3]
```diff
@@ -1,19 +1,107 @@
...
+    assert pfs_address, msg
+    if pfs_address == 'auto':
+        assert service_registry
```
Comment (ulope) [inline raiden/network/pathfinding.py]: Should have a message

File: raiden/network/pathfinding.py [hunk 4]
```diff
@@ -1,19 +1,107 @@
...
+        pfs_address = get_random_service(service_registry)
+        if pfs_address is None:
+            click.secho(
+                'The service registry has no registered PFS services and we dont '
```
Comment (ulope) [inline raiden/network/pathfinding.py]: More nitpicks: 
- `dont` -> `don't`
- PFS service (service service ;))

File: raiden/tests/integration/fixtures/smartcontracts.py [hunk 5]
```diff
@@ -108,6 +137,15 @@ def secret_registry_proxy(deploy_client, secret_registry_address, contract_manag
     )
 
 
+@pytest.fixture
+def service_registry_proxy(deploy_client, service_registry_address, contract_manager):
```
Comment (ulope) [inline raiden/tests/integration/fixtures/smartcontracts.py]: This seems to be unused.

File: raiden/tests/integration/fixtures/smartcontracts.py [hunk 6]
```diff
@@ -108,6 +137,15 @@ def secret_registry_proxy(deploy_client, secret_registry_address, contract_manag
...
+    return ServiceRegistry(
+        jsonrpc_client=deploy_client,
+        service_registry_address=to_canonical_address(service_registry_address),
```
Comment (ulope) [inline raiden/tests/integration/fixtures/smartcontracts.py]: IIUC `service_registry_address` can be `None` with will cause an exception in `to_canonical_address`.

File: raiden/tests/integration/network/proxies/test_service_registry.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        contract_manager=contract_manager,
+        service_registry_address=service_registry_address,
+    )
+    assert c1_service_proxy.service_count('latest') == 3
+
+    # Test that getting the url for each service address works
+    for idx, address in enumerate(addresses):
+        assert c1_service_proxy.get_service_url('latest', address) == urls[idx]
+    # Test that getting the url for a non-existing service address returns None
+    assert not c1_service_proxy.get_service_url('latest', to_checksum_address(HOP1))
```
Comment (ulope) [inline raiden/tests/integration/network/proxies/test_service_registry.py]: Nitpick: `assert ... is None` would be a bit more explicit

File: raiden/tests/integration/network/proxies/test_service_registry.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        'operator': 'John Doe',
+        'version': '0.0.1',
+    }
+
+    response = Mock()
+    response.configure_mock(status_code=200)
+    response.json = Mock(return_value=json_data)
+
+    # With basic routing configure pfs should return None
+    assert not configure_pfs(
```
Comment (ulope) [inline raiden/tests/integration/network/proxies/test_service_registry.py]: Same as above.

File: raiden/ui/app.py [hunk 9]
```diff
@@ -363,6 +341,11 @@ def run_app(
         secret_registry_contract_address is not None and
         endpoint_registry_contract_address is not None
     )
+    if environment_type == Environment.DEVELOPMENT:
+        contract_addresses_given = (
```
Comment (ulope) [inline raiden/ui/app.py]: Shouldn't this check also depend on the routing mode? If basic routing mode is used there is no need to know the service registry address.

File: raiden/ui/app.py [hunk 10]
```diff
@@ -426,6 +409,29 @@ def run_app(
         except AddressWrongContract:
             handle_contract_wrong_address('user_deposit', user_deposit_contract_address)
 
+    service_registry = None
+    if environment_type == Environment.DEVELOPMENT:
+        try:
```
Comment (ulope) [inline raiden/ui/app.py]: Same question as above. When basic routing is used we don't need to perform the check IMO.

File: raiden/ui/cli.py [hunk 11]
```diff
@@ -235,8 +252,11 @@ def options(func):
                 '--pathfinding-service-address',
                 help=(
                     'URL to the Raiden path finding service to request paths from.\n'
-                    'Example: https://pfs-ropsten.services-dev.raiden.network'
+                    'Example: https://pfs-ropsten.services-dev.raiden.network\n'
+                    'Can also be given the "auto" value so that raiden chooses a '
+                    'PFS randomly from the service registry address'
```
Comment (ulope) [inline raiden/ui/cli.py]: `address` -> `contract`?

File: raiden/network/pathfinding.py [hunk 12]
```diff
@@ -1,19 +1,107 @@
...
+
 
 def get_pfs_info(url: str) -> typing.Optional[typing.Dict]:
     try:
-        return requests.get(f'{url}/api/v1/info', timeout=DEFAULT_HTTP_REQUEST_TIMEOUT).json()
+        response = requests.get(
+            f'{url}/api/v1/info',
+            timeout=DEFAULT_HTTP_REQUEST_TIMEOUT,
+        )
+        return response.json()
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply raiden/network/pathfinding.py]: I just moved this function around, but sure I can add the proper exception handling here too

File: raiden/ui/app.py (hunk 9 above)
Comment (LefterisJP) [reply raiden/ui/app.py]: Well at the moment the two are tied together. But yes I guess this should depend on the routing mode in the long run.

File: raiden/network/proxies/service_registry.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+            self,
+            block_identifier: BlockSpecification,
+            index: int,
+    ) -> Optional[AddressHex]:
+        """Gets the address of a service by index. If index is out of range return None"""
+        try:
+            result = self.proxy.contract.functions.service_addresses(index).call(
+                block_identifier=block_identifier,
+            )
+        except web3.exceptions.BadFunctionCallOutput:
```
Comment (hackaugusto) [inline raiden/network/proxies/service_registry.py]: Why is this exception being handled here but no in the `get_service_url` call?

Does the `web3` library behave differently depending on the type of the attribute?

```solidity
    mapping(address => string) public urls;  // URLs of services for HTTP access
    address[] public service_addresses; // list of available services (ethereum addresses)
```

Comment (LefterisJP) [reply raiden/network/proxies/service_registry.py]: Yes. unknown key mapping to string will always return `''`.
But index out of range apparently throws BadFunctionCallOutput. There is a also a test for this.

**Pull request merged**